### ISSUE TITLE
Eager mode model.save will not save layer vars if the layer is in list of list

### ISSUE THREAD

**Issue by chenghuige** [tensorflow/tensorflow#22140]
tf 1.10.1 
like below model.save will not save CuDNNGRU vars if you use MyLayer in a tf.keras.Model.

    class MyLayer(keras.Model):
       def __init__(self):
        super(MyLayer, self).__init__()
          self.encodes =  [(keras.layers.CuDNNGRU(units=100, 
                                        return_sequences=True, 
                                        return_state=False, 
                                        recurrent_initializer='glorot_uniform'),)]

      def call(self, x):
        return self.encodes[0][0](x)

**Comment by allenlavoie** [comment]
A list works and a list of lists works but a list containing a tuple does not, correct? Tracking tuples would be nice but isn't being actively worked on. Happy to advise/review PRs.

**Comment by chenghuige** [comment]
@allenlavoie I think it would be nice to track and record any train able layer variables, not matter what data structure they are in.
I'm also interested in the mechanism of save vars.
1. Seems if use another layer like B in Layer A(B is a member of A), then we do not save any vars of B.
2. Then in order to save vars of B I have to make A a Model instead of a Layer, but for Model seems 
    we can not use add_weight, then I have to wrap members using add_weight using Layer.. A bit complex.

**Comment by allenlavoie** [comment]
Right, Layer doesn't have a __setattr__ override. Originally they were going to, but just defining it is a performance issue for functional layers (which create Layer objects in a loop and share variables by some other means).

You can add variables to a Model by assigning them to an attribute just like with sub-Layers. `add_weight` should work too; what's the issue?

**Comment by chenghuige** [comment]
@allenlavoie How to add variables to a Model by assigning them to an attribute just like with sub-Layers ?  I could not find any related examples.  Something like self.abc = self.add_variable("abc", [1, 100], initializer=tf.ones_initializer(dtype=tf.float32)) in Model.__init__() will fail and error say '`add_variable` is not supported on Networks. However, you may assign variables to attributes and they will show up in the weights and variables properties'  
Another question is if I want to use a Dense layer in a Model,  but the Dense layer out dim is unknown until I see the first input example, see input is [batch_size, dim] then I will use Dense(dim) , if using Layer
I think I can do this using build method, for Model how to do this ? I need to write another Layer to do this first ?

**Comment by allenlavoie** [comment]
I'd use `tf.Variable`:

```
>>> m = tf.keras.Model()
>>> m.v = tf.Variable(1.)
>>> m.trainable_variables
[<tf.Variable 'Variable:0' shape=() dtype=float32, numpy=1.0>]
```

It's fine to create new Layers in your call() method for subclassed Models. Just make sure they only get created on the first call. I think you can also define a build() method for Models if you'd like.

**Comment by chenghuige** [comment]
@allenlavoie I see, thanks!

**Comment by drpngx** [comment]
@chenghuige ok to close?

**Comment by chenghuige** [comment]
@drpngx  Fine close this.
